packages/backend/src/Tabulators/IRV.ts

```diff
@@ -23,94 +42,190 @@ export function IRV(candidates: string[], votes: ballot[], nWinners = 1, randomT
     }
 
+    
+
     let remainingCandidates = [...summaryData.candidates]
-    let activeVotes = parsedData.scores
-    let exhaustedVotes: ballot[] = []
+    let activeVotes: ballot[] = parsedData.scores
+    let exhaustedVotes: weightedVote[] = []
     let exhaustedVoteCount = 0
     let overVoteCount = 0
-    while (results.elected.length < 1) { //TODO: multiwinner STV
-        let roundVoteCounts = Array(summaryData.candidates.length).fill(0)
-        for (let i = activeVotes.length - 1; i >= 0; i--) {
-            // loop backwards over the ballots so exhausted votes can be removed
-            let ballot = activeVotes[i]
-            let topRemainingRank: number = 0
-            let topRemainingRankIndex: number = 0
-            let isOverVote = false
-            remainingCandidates.forEach(candidate => {
-                // loop over remaining candidates
-                // find top (non-zero) rank, if duplicate top ranks mark as overvote
-                if (ballot[candidate.index] == 0) return;
-                if (ballot[candidate.index] == null) return;
-
-                // candidate has rank
-
-                if (topRemainingRank == 0) {
-                    // set initial top rank
-                    topRemainingRankIndex = candidate.index
-                    topRemainingRank = ballot[candidate.index]
-                    isOverVote = false
-                }
-                else if (ballot[candidate.index] < ballot[topRemainingRankIndex]) {
-                    // set new top rank
-                    topRemainingRankIndex = candidate.index
-                    topRemainingRank = ballot[candidate.index]
-                    isOverVote = false
-                }
-                else if (ballot[candidate.index] === ballot[topRemainingRankIndex]) {
-                    // multiple top ranks, mark as over vote
-                    isOverVote = true
-                }
-            })
-            if (isOverVote) {
-                // ballot is overvote
-                overVoteCount += 1
-            }
-            if (topRemainingRank === 0 || isOverVote) {
-                // ballot is exhausted
-                exhaustedVoteCount += 1
-                exhaustedVotes.push(...activeVotes.splice(i, 1))
+
+    let weightedVotes: weightedVote[] = activeVotes.map(vote => ({ weight: Fraction(1), vote: vote, overvote: false }))
+
+    // Initialize candidate vote pools to empty arrays
+    let candidateVotes: weightedVote[][] = Array(summaryData.candidates.length)
+    for (var i = 0; i < candidateVotes.length; i++) {
+        candidateVotes[i] = [];
+    }
+
+    // Initial vote distribution, moves weighted votes into the appropriate candidate pools 
+    distributeVotes(remainingCandidates, candidateVotes, exhaustedVotes, weightedVotes)
+
+    // Set quota based on number of winners and if its proportional
+    let quota = 0
+    if (proportional) {
+        quota = Math.floor(activeVotes.length/(nWinners + 1) + 1)
```

Comment:
How do quota rules work when there's polarized factions? If there's 4 polarized factions fighting over 2 seats, then is it possible that no one reaches the quota?

Edit: I see that the non-proportional case recalculates the quota to account for this, but I feel like this is necessary for STV as well

Edit2: now I see we have a section where we elect the remaining candidates if remainingCandidates <= remainingSeats

Reply:
Yup, we will eliminate candidates until one gets above the quota or we get enough candidates to fill all seats.